### PR TITLE
test: circular_queue tests

### PR DESCRIPTION
**Pull request by mond77**

### CODE REVIEW

File: src/rpc/connection.rs [hunk 1]
```diff
@@ -693,3 +693,142 @@ impl CircularQueue {
...
+        queue.init();
+        let mut recv_meta_data = vec![];
+        let mut recv_data = vec![0u8; 1024];
+        let result = queue.register_callback(&mut recv_meta_data, &mut recv_data);
+        match result {
+            Ok(id) => {
+                let callback = queue.callbacks[id as usize];
+                unsafe {
+                    let oc = &*(callback as *mut OperationCallback);
+                    std::thread::spawn(|| match oc.channel.0.send(()) {
```
Comment (luanshaotong) [inline src/rpc/connection.rs:785]: If this new thread necessary?

Comment (luanshaotong) [reply src/rpc/connection.rs:785]: How about send signal in the same thread?

File: src/rpc/connection.rs [hunk 2]
```diff
@@ -693,3 +693,142 @@ impl CircularQueue {
...
+                                CallbackState::Done => assert!(true),
+                                _ => assert!(false),
+                            };
+                            assert_eq!(oc.meta_data_length, 24);
+                            assert_eq!(oc.data_length, 512);
+                        }
+                        Err(_) => assert!(false),
+                    });
+                }
+                match queue.response(id, 0, 24 as usize, 512 as usize) {
```
Comment (luanshaotong) [inline src/rpc/connection.rs:826]: You should not test only callability of the methods, but also correctness.

Comment (luanshaotong) [reply src/rpc/connection.rs:826]: Checking for predictable changes in the data structure may help.

**Review by luanshaotong — CHANGES_REQUESTED** [review]

File: src/rpc/connection.rs (hunk 1 above)
Comment (mond77) [reply src/rpc/connection.rs:785]: I thought maps::channel might be synchronized......

**Review by luanshaotong — APPROVED** [review]
LGTM

**Pull request merged**